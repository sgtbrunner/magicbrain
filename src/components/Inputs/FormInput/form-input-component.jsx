import React from 'react';
import PropTypes from 'prop-types';

const capitalizeFirstLetter = (string) => string[0].toUpperCase() + string.slice(1);

const FormInput = ({ name, type, errorData, onChange, onBlur, errorClass, isTopInput }) => (
  <div className={!isTopInput ? 'mt4' : ''}>
    <label className="db fw6 lh-copy mv2" htmlFor={name}>
      {capitalizeFirstLetter(name)}
      <input
        className={`pa2 mb0 ba bg-white w-100 ${errorClass}`}
        type={type}
        name={name}
        id={name}
        onChange={(event) => onChange(name, event)}
        onBlur={() => onBlur(name)}
      />
      {errorData.showError && (
        <p className="red f6 absolute mv0 right-0 left-0">{errorData.errorText}</p>
      )}
    </label>
  </div>
);

FormInput.propTypes = {
  name: PropTypes.string.isRequired,
  type: PropTypes.string,
  errorData: PropTypes.shape({
    showError: PropTypes.bool,
    errorText: PropTypes.string,
  }).isRequired,
  onChange: PropTypes.func,
  onBlur: PropTypes.func,
  errorClass: PropTypes.string,
  isTopInput: PropTypes.bool,
};

FormInput.defaultProps = {
  type: 'text',
  onChange: () => {},
  onBlur: () => {},
  errorClass: '',
  isTopInput: false,
};

export default FormInput;
