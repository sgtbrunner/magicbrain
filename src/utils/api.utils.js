const { REACT_APP_API_ENDPOINT } = process.env;

const fetchServer = ({
  path,
  method = 'POST',
  headers = { 'Content-Type': 'application/json' },
  body,
}) =>
  fetch(`${REACT_APP_API_ENDPOINT}${path}`, {
    method,
    headers,
    body: JSON.stringify({ ...body }),
  });

export default {
  signInUser: ({ email, password }) =>
    fetchServer({ path: '/signin', body: { email, password } }).then((response) => response.json()),

  registerUser: ({ name, email, password }) =>
    fetchServer({ path: '/register', body: { name, email, password } }).then((response) =>
      response.json()
    ),

  getImageBoundary: ({ input }) =>
    fetchServer({ path: '/imageurl', body: { input } }).then((response) => response.json()),

  updateImageCount: ({ id }) =>
    fetchServer({ path: '/image', method: 'PUT', body: { id } }).then((response) =>
      response.json()
    ),
};
