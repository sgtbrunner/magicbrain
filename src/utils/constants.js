export const PARTICLE_OPTIONS = {
  particles: {
    number: {
      value: 50,
      density: {
        enable: true,
        value_area: 500,
      },
    },
  },
};

export const { REACT_APP_API_ENDPOINT } = process.env;
