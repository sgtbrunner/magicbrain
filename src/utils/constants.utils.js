const PARTICLE_OPTIONS = {
  particles: {
    number: {
      value: 50,
      density: {
        enable: true,
        value_area: 500,
      },
    },
  },
};

export default PARTICLE_OPTIONS;
