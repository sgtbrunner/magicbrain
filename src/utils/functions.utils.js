export const getUserFromLocalStorage = () => localStorage.getItem('user');
