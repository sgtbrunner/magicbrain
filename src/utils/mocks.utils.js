export const USER = {
  id: 3,
  name: 'test',
  email: 'test@test.com',
  entries: 7,
};

export const OUTPUTS = {
  outputs: [
    {
      data: {
        regions: [
          {
            region_info: {
              bounding_box: {
                top_row: 0.5,
                left_col: 0.5,
                bottom_row: 0.5,
                right_col: 0.5,
              },
            },
          },
        ],
      },
    },
  ],
};

export const CREDENTIAL_ERROR = {
  error: 'You have entered an invalid username and/or password',
};

export const SERVER_ERROR = {
  error: 'Something went wrong. Please try again later.',
};
