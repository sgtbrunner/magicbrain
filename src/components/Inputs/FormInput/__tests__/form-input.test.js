import { screen } from '@testing-library/react';

import { capitalizeFirstLetter } from '../../../../utils/functions.utils';
import { mountComponent } from '../../../../utils/test.utils';
import FormInput from '..';

const TEST = 'test';
const ERROR = 'Ooopps...an error occurred';
const PASSWORD = 'password';
const EMAIL = 'email';
const ALERT = 'alert';
const FORM_INPUT = 'form-input';
const SPACING_CLASS = 'mt4';

const emptyErrorData = {};

const mountFormInputComponent = (props) => mountComponent({ component: FormInput, props });

describe('FormInput component', () => {
  it('should match snapshot when input HAS NO errors', () => {
    const { container } = mountFormInputComponent({
      name: TEST,
      errorData: { errorText: ERROR, showError: false },
    });

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when input HAS errors', () => {
    const { container } = mountFormInputComponent({
      name: TEST,
      errorData: { errorText: ERROR, showError: true },
    });

    expect(container).toMatchSnapshot();
  });

  it(`should render a 'Email' input field when name is 'email'`, () => {
    mountFormInputComponent({ name: EMAIL, errorData: emptyErrorData });

    const label = screen.getByLabelText(capitalizeFirstLetter(EMAIL));
    expect(label).toBeInTheDocument();
  });

  it(`should render a 'Name' input field when name is 'password'`, () => {
    mountFormInputComponent({ name: PASSWORD, errorData: emptyErrorData });

    const label = screen.getByLabelText(capitalizeFirstLetter(PASSWORD));
    expect(label).toBeInTheDocument();
  });

  it(`should HAVE top spacing when component 'isTopInput' prop is false`, () => {
    mountFormInputComponent({ name: TEST, errorData: emptyErrorData, isTopInput: false });

    const input = screen.getByTestId(FORM_INPUT);
    expect(input).toHaveClass(SPACING_CLASS);
  });

  it(`should NOT HAVE top spacing when component 'isTopInput' prop is true`, () => {
    mountFormInputComponent({ name: TEST, errorData: emptyErrorData, isTopInput: true });

    const input = screen.getByTestId(FORM_INPUT);
    expect(input).not.toHaveClass(SPACING_CLASS);
  });

  it(`should render error message when errorData HAS error`, () => {
    mountFormInputComponent({ name: TEST, errorData: { errorText: ERROR, showError: true } });

    const error = screen.getByRole(ALERT);
    expect(error).toBeInTheDocument();
  });

  it(`should NOT render error message when errorData HAS NO error`, () => {
    mountFormInputComponent({ name: TEST, errorData: { errorText: ERROR, showError: false } });

    const error = screen.queryByRole(ALERT);
    expect(error).not.toBeInTheDocument();
  });
});
