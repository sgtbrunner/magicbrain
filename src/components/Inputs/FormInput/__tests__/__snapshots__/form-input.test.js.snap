// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput component should match snapshot when input HAS NO errors 1`] = `
<div>
  <div
    class="mt4"
    data-testid="form-input"
  >
    <label
      class="db fw6 lh-copy mv2"
      for="test"
    >
      Test
      <input
        class="pa2 mb0 ba bg-white w-100 br2 () => showError ? 'b--red' : ''"
        id="test"
        name="test"
        type="text"
      />
    </label>
  </div>
</div>
`;

exports[`FormInput component should match snapshot when input HAS errors 1`] = `
<div>
  <div
    class="mt4"
    data-testid="form-input"
  >
    <label
      class="db fw6 lh-copy mv2"
      for="test"
    >
      Test
      <input
        class="pa2 mb0 ba bg-white w-100 br2 () => showError ? 'b--red' : ''"
        id="test"
        name="test"
        type="text"
      />
      <p
        class="red f6 absolute mv0 right-0 left-0"
        role="alert"
      >
        Ooopps...an error occurred
      </p>
    </label>
  </div>
</div>
`;
